### PR TITLE
Normalize Java SDK (JDK) location logic across the tool 

### LINKED ISSUE #122609: JDK actually used for Gradle doesn't appear in `flutter doctor -v` nor in `flutter run -v`
Update: see comments starting at https://github.com/flutter/flutter/issues/122609#issuecomment-1474273992 . Original report follows.

---

After Android configuration changes (creating a new project, adding plugins) I get this error. I am using the latest Android Studio Preview and stable gradle versions. The only way I've found to fix this is to open the Android native part of the project in Android Studio and click the gradle sync button. Either the root issue should be resolved or the Flutter tool should be doing whatever the gradle sync button does automatically.

Related to https://github.com/flutter/flutter/issues/122376

When I try to run a fresh project on an Android device:
```console
Running "flutter pub get" in gradle_test...
Resolving dependencies...
  async 2.10.0 (2.11.0 available)
  characters 1.2.1 (1.3.0 available)
  collection 1.17.0 (1.17.1 available)
  js 0.6.5 (0.6.7 available)
  matcher 0.12.13 (0.12.14 available)
  material_color_utilities 0.2.0 (0.3.0 available)
  meta 1.8.0 (1.9.0 available)
  path 1.8.2 (1.8.3 available)
  test_api 0.4.16 (0.4.18 available)
Got dependencies!
Launching lib/main.dart on sdk gphone64 arm64 in debug mode...
main.dart:1

FAILURE: Build failed with an exception.

* Where:
Build file '/Users/rexios/repos/gradle_test/android/build.gradle' line: 26

* What went wrong:
A problem occurred evaluating root project 'android'.
> A problem occurred configuring project ':app'.
   > Could not open proj generic class cache for build file '/Users/rexios/repos/gradle_test/android/app/build.gradle' (/Users/rexios/.gradle/caches/7.5/scripts/up9lbpef9oco35z135w5mojx).
      > BUG! exception in phase 'semantic analysis' in source unit '_BuildScript_' Unsupported class file major version 63

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 12s
Exception: Gradle task assembleDebug failed with exit code 1
Exited
```

### PR DESCRIPTION
**Pull request by andrewkolos**
Fixes #122609.

Related: https://github.com/flutter/flutter/issues/122376

## Context 
`flutter doctor -v` prints the path of the `java` binary that one might assume to be used for all Android-related features (building apps using Gradle, managing emulators, etc.). However, this `java`-locating logic is different than that used specifically when building Android apps using Gradle. See discussions on #122609 for more info.

To summarize:

For most activities, including `flutter doctor` validators and Android SDK-related features (creating emulators, checking license acceptance, etc..), we use the following logic:

1. use the JDK bundled with Android Studio if we can find it; otherwise
2. use `JAVA_HOME` (from env) if it is set; otherwise
3. (on macOS only) use the path returned by `/usr/libexec/java_home -v 1.8` if any; otherwise
4. use `which java` (search for `java` on PATH)

However, specifically when building Android apps using Gradle, `/usr/libexec/java_home` (step 3) will never be consulted. Instead, we'll explicitly set `JAVA_HOME` to the Android Studio-bundled JDK if it exists; otherwise, `gradlew` will implicitly consult `JAVA_HOME` and then look for `java` on PATH.

This has the potential to be confusing to those debugging build issues in Android-enabled flutter projects.

## Change
We'll stop using `/usr/libexec/java_home -v 1.8` as a second-to-last fallback for non-Gradle build activities.

## Pre-launch Checklist
<details>

<summary>Pre-launch Checklist</summary>

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
</details>

### CODE REVIEW

File: packages/flutter_tools/lib/src/android/android_sdk.dart [hunk 1]
```diff
@@ -426,22 +426,6 @@ class AndroidSdk {
       return fileSystem.path.join(javaHomeEnv, 'bin', 'java');
     }
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/android/android_sdk.dart:428]: Looks like we have no unit tests for this function. Could you add tests that validate we discover the right path in the situations where:

1. it's present in Android Studio
2. not present in AS but JAVA_HOME is set
3. neither 1 nor 2, but `which java` returns a valid value
4. they really don't have any java binaries we can find

**Comment by andrewkolos** [comment]
This is (technically) a breaking change since the version of Java found by `flutter doctor` and used by `flutter` to perform Android SDK-related operations[^1] may change for macOS users for which 1) the flutter tool cannot find the Android Studio-bundled JDK, 2) the `JAVA_HOME` env variable is unset, and 3) `usr/libexec/java_home -v 1.8` returns something different than `which java`. However, I have to imagine this situation is pretty unlikely.

Unrelated FYI for @reidbaker: this change shouldn't block any other work surrounding https://github.com/flutter/flutter/issues/122376. I'll save any refactoring for once the critical work is done.

[^1]: (validating Android SDK license acceptance, managing emulators, etc.)

**Comment by reidbaker** [comment]
Solid this will get a review today as soon as I am out of meetings

File: packages/flutter_tools/lib/src/android/android_sdk.dart (hunk 1 above)
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/android/android_sdk.dart:428]: done

File: packages/flutter_tools/lib/src/android/android_sdk.dart [hunk 2]
```diff
@@ -426,22 +426,6 @@ class AndroidSdk {
...
```
Comment (gnprice) [inline packages/flutter_tools/lib/src/android/android_sdk.dart:427]: It'd be good to add a comment at the top of this function saying it needs to be kept in sync with how `android/gradle.dart` sets `JAVA_HOME` when running Gradle. Also a comment at the relevant places in that file (the two places that https://github.com/flutter/flutter/pull/124152/files#diff-3c44d2e8279ea8d917b4fb942252d4649689154c36cf74227a1a2a879aaf1915 changes to use `_javaSdkHome`) saying they need to be kept in sync with what `findJavaBinary` does.

**Review by gnprice — COMMENTED** [review]
Looks good to me.

I think a refactor like #124152 is important in order to prevent the two implementations of which JDK to use from diverging again in the future. But this fix makes sense as a way of reconciling the behavior for now while minimizing conflicts with related urgent changes.

In addition to filing that tech-debt issue, I'd suggest adding a couple of comments about the need to keep these separate implementations in sync.

**Review by christopherfujino — COMMENTED** [review]
LGTM, but I'll leave final approval to @reidbaker

File: packages/flutter_tools/lib/src/android/android_sdk.dart (hunk 1 above)
Comment (reidbaker) [reply packages/flutter_tools/lib/src/android/android_sdk.dart:494]: FWIW I think my pr #123916 might have added some tests.

**Review by reidbaker — APPROVED** [review]

File: packages/flutter_tools/lib/src/android/android_sdk.dart (hunk 2 above)
Comment (reidbaker) [reply packages/flutter_tools/lib/src/android/android_sdk.dart:493]: @andrewkolos I agree with @gnprice here.

**Pull request merged**